[Week5 Spring] 고양이 장난감가게 유효성 추가,  회원 관리 만들기

mikekang47 opened the pull request:
- [x] 고양이 장난감 목록 얻기 - GET /products

- [x] 고양이 장난감 상세 조회하기 - GET /products/{id}

- [x] 고양이 장난감 등록하기 - POST /products

- [x] 고양이 장난감 수정하기 - PATCH /products/{id}

- [x] 고양이 장난감 삭제하기 - DELETE /products/{id}



- [x] 회원 생성하기 - POST /user

- [x] 회원 수정하기 - PATCH /user/{id}

- [x] 회원 삭제하기 - DELETE /user/{id}


dozer mapper를 활용해서 java bean 과 java bean을 연결한다.

mikekang47 committed:
Add @CrossOrigin to connet backend and frontend
Add @DisplayName in ProductControllerTest for easy to check out

mikekang47 committed:
Add dozerMapper in App.java and build.gradle

johngrib commented on app/build.gradle:
```diff
@@ -51,6 +57,9 @@ dependencies {
     annotationProcessor 'org.projectlombok:lombok:1.18.16'
 
+    //DozerMapper
+    implementation 'com.github.dozermapper:dozer-core:6.4.0'
```
`DozerMapper`를 리뷰어에게 소개하고, 이 라이브러리를 도입하자고 설득하는 짧은 글을 PR 본문에 작성해 보세요.

johngrib reviewed (commented):
경록님 PR을 올려주셨군요. 화이팅!

mikekang47 committed:
Add Mapper to createProudct in ProductService

mikekang47 committed:
Add dozermapper in ProductService constructor and its test

mikekang47 committed:
Make updateProduct to use mapper and modified its test
Added changeWith

mikekang47 committed:
Make User class and UserTest
Using builder Pattern by Lombok @Builder

mikekang47 committed:
Make UserService, JpaUserRepository, UserRepository and theirs test

mikekang47 committed:
Make UserController and its test

mikekang47 committed:
Added getUsers in UserServiceTest
And setting void setUp()

mikekang47 committed:
Make a case when there is no user in the list

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,25 @@
+
+@Service
+@Transactional
+public class UserService {
```
이 서비스에 동료 개발자를 위한 JavaDoc 주석을 작성해 주세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/UserController.java:
```diff
@@ -0,0 +1,26 @@
+@RestController
+@CrossOrigin
+@RequestMapping("/user")
+public class UserController {
```
여기에도 JavaDoc 주석을 작성해 주세요. 주요 서비스와 컨트롤러에는 꼭 작성해 주는 습관을 들입시다.

mikekang47 replied:
적용해보겠습니다~

johngrib commented on app/src/test/java/com/codesoom/assignment/application/ProductServiceTest.java:
```diff
@@ -75,7 +78,7 @@ void getProductWithExsitedId() {
     }
 
     @Test
-    void getProductWithNotExsitedId() {
+    void getProductWithNotExistedId() {
```
이런 오타가 있었군요 ㅎㅎ

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,56 @@
+
+        assertThat(user.getName()).isEqualTo("mikekang");
+    }
+}
```
파일 마지막에 개행 문자를 추가해 주세요.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    @Test
+    @DisplayName("저장된 회원이 있다면 회원 목록을 리턴한다")
```
이렇게 수정하면 어떨까요?
```suggestion
    @DisplayName("getUsers는 저장된 회원이 있다면 회원 목록을 리턴한다")
```

johngrib reviewed (commented):
`@Nested`를 사용한 계층 구조의 테스트 코드도 고려해 보세요.

mikekang47 commented:
> `@Nested`를 사용한 계층 구조의 테스트 코드도 고려해 보세요.

공부해서 사용해보도록 하겠습니다!

mikekang47 committed:
Modified the things that was reviewed
Add list() in UserContrllerTest

mikekang47 committed:
Make UserNotfoundException class
Plus added getuserWithExistedId in UserServiceTest
Plus added detailWithExistedUser in UserControllerTest and Usercontroller

mikekang47 committed:
Make UserData class
Plus make createUser test in UserServiceTest.java

mikekang47 committed:
Make create in UserController and its test

mikekang47 committed:
Add detailWithNotExistedUser
For to do so, added userException in erroradvice class

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -8,6 +8,9 @@
 import javax.transaction.Transactional;
 import java.util.List;
 
+/**
+ * 회원 정보와 관련된 비지니스 로직을 처리
+ */
```
잘 작성하셨습니다. 그런데 class 를 설명하는 Summary는 완성된 문장으로 작성하는 관례가 있어서 마침표를 찍어 주시는 것이 좋습니다.
`회원 정보와 관련된 비지니스 로직을 처리합니다.`

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/UserController.java:
```diff
@@ -9,6 +9,9 @@
 
 import java.util.List;
 
+/**
+ * 사용자의 생성, 업데이트, 삭제를 처리
+ */
```
여기도요. `사용자의 생성, 업데이트, 삭제를 처리합니다.`

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/UserRepository.java:
```diff
@@ -1,9 +1,11 @@
-
     List<User> findAll();
 
+    Optional<User> findById(Long id);
```
탁월한 선택입니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/ProductServiceTest.java:
```diff
@@ -70,7 +70,7 @@ void getProducts() {
     }
 
     @Test
-    void getProductWithExsitedId() {
+    void getProductWithExistedId() {
```
이런 오타가 있었군요!

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -54,4 +57,14 @@ void getUsers() {
+    @Test
+    @DisplayName("getUserWithExistedId 는 존재하는 특정 회원을 리턴한다.")
+    void getUserWithExistedId() {
+        User user = userService.getUser(1L);
```
`@DisplayName`의 설명과 테스트 코드 내용이 다른 것 같아요.

johngrib commented on app/src/main/java/com/codesoom/assignment/dto/UserData.java:
```diff
@@ -0,0 +1,28 @@
+
+import javax.validation.constraints.NotBlank;
+
+@Setter
```
여기에 `@Setter`가 있으면 `id`에도 setter 메소드가 생기게 됩니다.

mikekang47 committed:
Implemented update and deleteUser and theirs tests

mikekang47 committed:
Refactored about getter and setter

mikekang47 committed:
Refactored about app.test

mikekang47 committed:
Added createWithInvalidAttributes
Refactoring

mikekang47 committed:
Add setter in ProududctData to pass the test

mikekang47 committed:
Refactoring

mikekang47 commented:
User 관련한 테스트는 잘 동작하는데, 왜 productControllertest에서 create와 update관련한 메서드들이 자꾸 터질까요....ㅜㅜ
처음과 비교해봐도 달라진점이 없는데 어디에서 문제가 발생하는지 도무지 찾을 수가 없네요ㅜ

johngrib commented:
> User 관련한 테스트는 잘 동작하는데, 왜 productControllertest에서 create와 update관련한 메서드들이 자꾸 터질까요....ㅜㅜ
> 처음과 비교해봐도 달라진점이 없는데 어디에서 문제가 발생하는지 도무지 찾을 수가 없네요ㅜ

제가 받아서 돌려보니 다음 부분이 문제입니다.

![image](https://user-images.githubusercontent.com/1855714/109419201-a9fd9b00-7a0f-11eb-9161-72bfeb9b0dcc.png)

`ProductData.imageUrl`에 `@NotNull`을 붙여 두셨는데요, `ProductController`에 가보면 이렇게 `@Valid`가 붙어 있습니다.
![image](https://user-images.githubusercontent.com/1855714/109419235-cf8aa480-7a0f-11eb-8b13-f04f03f1ef10.png)

즉, 테스트 코드에서 `imageUrl`을 제공하지 않아서 컨트롤러 `@Valid` 검사에서 튕겨 내는 것입니다. 테스트 코드에서 이렇게 `imageUrl`을 추가해보세요.

![image](https://user-images.githubusercontent.com/1855714/109419257-ef21cd00-7a0f-11eb-9887-9ed339d884b4.png)

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/User.java:
```diff
@@ -1,17 +1,17 @@
 
 @Getter
-@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Setter
```
클래스에 붙이는 `@Setter`는 신중하게 사용해야 합니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -123,19 +123,17 @@ void create() throws Exception {
+    void createWithInvalidAttributes() throws Exception {
+        mockMvc.perform(
+                post("/user")
+                        .accept(MediaType.APPLICATION_JSON_UTF8)
```
이건 deprecated 되었습니다. `MediaType` 클래스에 들어가서 JavaDoc 주석을 읽어보시고 이걸 무엇으로 바꿔야 하는지 알아보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Product.java:
```diff
@@ -1,9 +1,6 @@
-import lombok.Builder;
-import lombok.Getter;
-import lombok.NoArgsConstructor;
+import lombok.*;
```
`import *`의 사용은 지양하는 것이 좋습니다. 좋지 않은 습관입니다.

johngrib reviewed (commented):
한 주간 고생 많으셨습니다. 주간 회고를 작성해 공유해주세요.

mikekang47 committed:
Refactored problems

mikekang47 commented:
> > User 관련한 테스트는 잘 동작하는데, 왜 productControllertest에서 create와 update관련한 메서드들이 자꾸 터질까요....ㅜㅜ
> > 처음과 비교해봐도 달라진점이 없는데 어디에서 문제가 발생하는지 도무지 찾을 수가 없네요ㅜ
> 
> 제가 받아서 돌려보니 다음 부분이 문제입니다.
> 
> ![image](https://user-images.githubusercontent.com/1855714/109419201-a9fd9b00-7a0f-11eb-9161-72bfeb9b0dcc.png)
> 
> `ProductData.imageUrl`에 `@NotNull`을 붙여 두셨는데요, `ProductController`에 가보면 이렇게 `@Valid`가 붙어 있습니다.
> ![image](https://user-images.githubusercontent.com/1855714/109419235-cf8aa480-7a0f-11eb-8b13-f04f03f1ef10.png)
> 
> 즉, 테스트 코드에서 `imageUrl`을 제공하지 않아서 컨트롤러 `@Valid` 검사에서 튕겨 내는 것입니다. 테스트 코드에서 이렇게 `imageUrl`을 추가해보세요.
> 
> ![image](https://user-images.githubusercontent.com/1855714/109419257-ef21cd00-7a0f-11eb-9887-9ed339d884b4.png)

감사합니다!!!

mikekang47 committed:
Add putUser

mikekang47 committed:
Refacotored about productService

mikekang47 committed:
Refactoring userController and userService

mikekang47 committed:
add put method to pass the test

mikekang47 committed:
add put method in userServiceTest

mikekang47 committed:
Modify typo

mikekang47 committed:
install npm test

johngrib commented:
한 주간 고생 많으셨습니다.
